dockertest/subtestbase.py

```diff
@@ -8,10 +8,69 @@
 # pylint: disable=W0403
 
 import logging
+import os.path
+import sys
 import traceback
+from autotest.client import utils
 from xceptions import DockerTestFail
 from xceptions import DockerTestNAError
-from config import get_as_list
+from config import CONFIGCUSTOMS, get_as_list
+from docker_daemon import which_docker
+
+
+def docker_rpm():
+    """
+    Returns the full NVRA of the currently-installed docker or docker-latest.
+
+    FIXME: this won't work for container-engine. That's tricky, and
+    not high priority, so let's save it for a subsequent PR.
+    """
+    return utils.run("rpm -q %s" % which_docker()).stdout.strip()
+
+
+def known_failures_file():
+    """
+    Returns path to a file containing a list of known failures.
+
+    Format of this file is:
+
+          <NVRA>    <subtest name>   <human-friendly description>
+
+    Fields are separated by whitespace; spaces in the third (description)
+    field are OK. This lets us column-align the fields for readability.
+    Comment lines (starting with '#') and blank lines are ignored.
+    """
+    return os.path.join(CONFIGCUSTOMS, 'known_failures.txt')
+
+
+def known_failures():
+    """
+    Returns a dict containing known test failures. Primary key is
+    subtest name (e.g. docker_cli/sub/subsub), value is another dict
+    whose key is docker NVRA (e.g. docker-1.12.5-8.el7.x86_64),
+    value of that is a string description of the problem (e.g.
+    a bz number and comment).
+    """
+    known_failures_path = known_failures_file()
+    known = {}
+    try:
+        known_failures_fh = open(known_failures_path, 'r')
+    except IOError, excpt:
+        SubBase.logwarning("Skipping known_failure check: %s" % excpt)
+        return known
+
+    for row in known_failures_fh:
+        row = row.strip()
+        if row and not row.startswith('#'):
+            try:
+                nvra, subtest, description = row.strip().split(None, 2)
+                if subtest not in known:
+                    known[subtest] = {}
+                known[subtest][nvra] = description
+            except ValueError:
+                SubBase.logwarning("Bad row in %s: %s"
+                                   % (known_failures_path, row))
+    return known
```

Comment:
Would it be a major PITA to move the above functions into the ``environment`` module?

That's really it's purpose (inspecting/doing stuff with the os-environment), moving them later certainly would be a PITA.

I can give it a go and send you a patch if you're disinclined (unless there's a compelling reason to leave them here).

Reply:
FWIW: My motivation is mainly about general code-orginization but ...wait a sec...no, I'm wrong.  Really the only function that should move to ``environment`` is ``docker_rpm()``.  The ``known_failure...`` functions are more specific to subtest than environment.  My bad.

Reply:
Whew. Okay, thanks. Moving `known_failures` is turning out more complicated than I had first thought. (And I agree, known_failures belongs closer to subtest stuff). Undoing.

Reply:
Okay, glanced at the unittests and code WRT ``docker_rpm()``.  Doesn't appear like it would be all that painful to move over to ``environment.docker_rpm()``.  Is that cool?

Reply:
Done.